### PR TITLE
[ETHEREUM-CONTRACTS/SDK-CORE] Add `send` operation to BatchCall

### LINKED ISSUE #1194: [ETHEREUM-CONTRACTS] New Superfluid BatchCall Operation: ERC777.send

### PR DESCRIPTION
**Pull request by 0xdavinchee**
Related to this [issue](https://github.com/superfluid-finance/protocol-monorepo/issues/1194) and this [comment on slack](https://superfluidhq.slack.com/archives/C02G67JCG06/p1670005689608459).

**note: the tests I've added so far are just happy path cases, but I think it would be good to spend some time to discuss and make sure that these changes are safe.**

**contracts**
- Added new batch operation definition in contracts
- Added `operationSend` function for `SuperToken.sol` only to be called by host
- Did some wild stuff to get `SuperfluidFrameworkDeployer` to be deployed (https://github.com/NomicFoundation/hardhat/issues/3404)
  - **my current solution was just to unblock me from running tests for sdk-core**

**sdk-core**
- added `send` to sdk-core
- added `batchCall` definitions to sdk-core

cc @d10r @ngmachado

### CODE REVIEW

File: packages/ethereum-contracts/contracts/interfaces/superfluid/Definitions.sol [hunk 1]
```diff
@@ -150,6 +150,15 @@ library BatchOperation {
      * )
      */
     uint32 constant internal OPERATION_TYPE_ERC20_TRANSFER_FROM = 2;
+    /**
+     * @dev ERC777.send batch operation type
+     *
+     * Call spec:
+     * ISuperToken(target).operationSend(
+     *     abi.decode(data, (address recipient, uint256 amount)
```
Comment (hellwolf) [inline packages/ethereum-contracts/contracts/interfaces/superfluid/Definitions.sol]: We are missing the ERC777 data.

File: packages/ethereum-contracts/contracts/interfaces/superfluid/ISuperToken.sol [hunk 2]
```diff
@@ -194,7 +194,7 @@ interface ISuperToken is ISuperfluidToken, TokenInfo, IERC20, IERC777 {
 
     /**
      * @dev Destroys `amount` tokens from the caller's account, reducing the
-     * total supply.
+     * total supply and transfers the underlying token to the caller's account.
```
Comment (hellwolf) [inline packages/ethereum-contracts/contracts/interfaces/superfluid/ISuperToken.sol:197]: :ok:

File: packages/ethereum-contracts/contracts/interfaces/superfluid/ISuperToken.sol [hunk 3]
```diff
@@ -454,6 +454,21 @@ interface ISuperToken is ISuperfluidToken, TokenInfo, IERC20, IERC777 {
...
+    * @param recipient The recipient of the funds.
+    * @param amount Number of tokens to be transferred.
+    *
+    * @custom:modifiers 
+    *  - onlyHost
+    */
+    function operationSend(
+        address spender,
+        address recipient,
+        uint256 amount
```
Comment (hellwolf) [inline packages/ethereum-contracts/contracts/interfaces/superfluid/ISuperToken.sol]: add the ERC777 data

File: packages/ethereum-contracts/contracts/utils/SuperfluidFrameworkDeployer.sol [hunk 4]
```diff
@@ -2,10 +2,6 @@
 pragma solidity ^0.8.0;
```
Comment (hellwolf) [inline packages/ethereum-contracts/contracts/utils/SuperfluidFrameworkDeployer.sol]: Separate PR for this: https://github.com/superfluid-finance/protocol-monorepo/issues/1220

**Review by hellwolf — DISMISSED** [review]
Everything else is reviewed and looks good, but it is blocked by https://github.com/superfluid-finance/protocol-monorepo/issues/1220 for now.

File: packages/ethereum-contracts/contracts/interfaces/superfluid/Definitions.sol (hunk 1 above)
Comment (0xdavinchee) [reply packages/ethereum-contracts/contracts/interfaces/superfluid/Definitions.sol]: added

File: packages/ethereum-contracts/contracts/interfaces/superfluid/ISuperToken.sol (hunk 3 above)
Comment (0xdavinchee) [reply packages/ethereum-contracts/contracts/interfaces/superfluid/ISuperToken.sol]: added

File: packages/ethereum-contracts/contracts/interfaces/superfluid/Definitions.sol [hunk 5]
```diff
@@ -174,7 +183,7 @@ library BatchOperation {
      * Call spec:
      * callAgreement(
      *     ISuperAgreement(target)),
-     *     abi.decode(data, (bytes calldata, bytes userdata)
+     *     abi.decode(data, (bytes callData, bytes userData)
```
Comment (0xdavinchee) [inline packages/ethereum-contracts/contracts/interfaces/superfluid/Definitions.sol:186]: modified the documentation here as `calldata` is a reserved keyword in solidity

File: packages/ethereum-contracts/contracts/interfaces/superfluid/ISuperToken.sol [hunk 6]
```diff
@@ -450,10 +450,10 @@ interface ISuperToken is ISuperfluidToken, TokenInfo, IERC20, IERC777 {
     ) external;
 
     /**
-    * @dev Perform ERC20 transfer from by host contract.
+    * @dev Perform ERC20 transferFrom by host contract.
     * @param account The account to spend sender's funds.
-    * @param spender  The account where the funds is sent from.
-    * @param recipient The recipient of thefunds.
+    * @param spender The account where the funds is sent from.
+    * @param recipient The recipient of the funds.
```
Comment (0xdavinchee) [inline packages/ethereum-contracts/contracts/interfaces/superfluid/ISuperToken.sol:456]: typo fix

File: packages/ethereum-contracts/contracts/superfluid/SuperToken.sol [hunk 7]
```diff
@@ -688,6 +688,18 @@ contract SuperToken is
...
+    function operationSend(
+        address spender,
+        address recipient,
+        uint256 amount,
+        bytes memory userData
+    ) 
+        external override
+        onlyHost
+    {
+        _send(msg.sender, spender, recipient, amount, userData, "", userData.length != 0);
```
Comment (0xdavinchee) [inline packages/ethereum-contracts/contracts/superfluid/SuperToken.sol]: we retain the `userData.length != 0` rule for `requireReceptionAck`

File: packages/ethereum-contracts/contracts/superfluid/Superfluid.sol [hunk 8]
```diff
@@ -796,14 +796,22 @@ contract Superfluid is
                     sender,
                     receiver,
                     amount);
+            } else if (operationType == BatchOperation.OPERATION_TYPE_ERC777_SEND) {
```
Comment (0xdavinchee) [inline packages/ethereum-contracts/contracts/superfluid/Superfluid.sol:799]: send support in `_batchCall`

File: packages/ethereum-contracts/test/contracts/superfluid/SuperToken.NonStandard.test.ts [hunk 9]
```diff
@@ -936,6 +936,11 @@ describe("SuperToken's Non Standard Functions", function () {
                 superToken,
                 "SF_TOKEN_ONLY_HOST"
             );
+            await expectCustomError(
```
Comment (0xdavinchee) [inline packages/ethereum-contracts/test/contracts/superfluid/SuperToken.NonStandard.test.ts:939]: test for `operationSend` reverting if called outside of host being the caller

**Review by 0xdavinchee — COMMENTED** [review]
some signposts

File: packages/ethereum-contracts/contracts/superfluid/SuperToken.sol (hunk 7 above)
Comment (hellwolf) [reply packages/ethereum-contracts/contracts/superfluid/SuperToken.sol]: should follow what send does

```
    function send(address recipient, uint256 amount, bytes calldata data) external override {
        _send(msg.sender, msg.sender, recipient, amount, data, "", true);
    }
```

Comment (0xdavinchee) [reply packages/ethereum-contracts/contracts/superfluid/SuperToken.sol]: done

**Review by hellwolf — APPROVED** [review]

**Pull request merged**